_posts/2017-08-18-Refactoring-with-reactJS-shared-state.md

```diff
@@ -0,0 +1,252 @@
+    }
+}
+{% endhighlight %}
+
+Using our `FailureBanner` example from last time, let's suppose that we want to display the failure banner to a user only after they've clicked a button on the page. Let's assume that the button is rendered by the server and just exists on the existing page.
+
+{% highlight html linenos %}
+<script type="text/javascript">
+    $('#failureButton').click(function(e) {
+        // LEGACY CODE - does some crazy stuff to the rest of the page
+
+        // Here is where we want to display our React failure banner with a specific message
+    });
+</script>
+
+<div data-react-component='FailureBanner' data-is-shown='false' data-message=''></div>
+<button id='failureButton' >Do The Thing</button>
+
+<!-- Include the react initialization code from last time
+It renders the FailureBanner on page load with the isShown and message props set to the values in the tag -->
+<script src='index.js'></script>
+{% endhighlight %}
+
+In this example, we already have an instance of the `FailureBanner` rendered into the page with some default properties. In this case, it's hidden from the user because we set the isShown prop to `false`. Once the user clicks our legacy failure button, the application does a bunch of stuff. At the very end, we've decided that it should now show a failure banner to the user. We want to update the `FailureBanner` component's props so that `isShown` is true and the `message` says "Sorry, we couldn't do the thing."
+
+# How do we update component props?
+
+One solution we’ve seen before is to unmount the React component, then re-render the component into the DOM with the updated props. It’s a workable solution, but we’ve always disliked it because it completely ignores one of the biggest benefits of using React: the [Virtual DOM](https://www.codecademy.com/articles/react-virtual-dom).
+
+Instead, we took a look around and decided that we needed an application state management solution. In our greenfield React project we use [Redux](http://redux.js.org/) to manage application state. Since we already had this knowledge on the team, we decided to use Redux in our legacy refactoring as well. One could probably just as easily use a reactive library like [RxJS](http://reactivex.io/rxjs/. We’ll have to leave that as an exercise for another day.
+
+Our Redux store will be the source of truth for the application's state. Our React components will be connected to the store and can dispatch changes to the store and/or react to application state changes appropriately. Similarly, our legacy Javascript will be able to access the store and dispatch actions to change application state.
+
+# Create the Redux Store
+
+First, we need to create a Redux store for our application. We're going to keep this piece with the React initilization script from part 1. Since the store needs to be accessible from all of our legacy code, we decided to create the store as a global. Please keep your 'boos' to a light rumble.
```

Comment:
Since the store needs to be accessible from all of our legacy code we decided to take the meaning of "global store" literally and attach it to the window object. Please keep your 'boos' to a light rumble and stick with us. We promise it gets better.